module/zcommon/zfs_fletcher.c

```diff
@@ -245,7 +214,107 @@ fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
 	ZIO_SET_CHECKSUM(zcp, a, b, c, d);
 }
 
+static const struct fletcher_4_calls fletcher_4_generic_calls = {
+	.init = fletcher_4_generic_init,
+	.compute = fletcher_4_generic,
+	.compute_byteswap = fletcher_4_generic_byteswap,
+	.name = "generic"
+};
+
+static const struct fletcher_4_calls *chosen = &fletcher_4_generic_calls;
+
+void
+fletcher_4_native(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	chosen->init(zcp);
+	chosen->compute(buf, size, zcp);
+	if (chosen->fini != NULL)
+		chosen->fini(zcp);
```

Comment:
I'd feel much better about this ->init ->compute ->fini model if we has some before and after performance numbers for comparison.